File: android/app/src/main/java/eu/agilmea/ik/tracker/Diary.java
package eu.agilmea.ik.tracker;

import android.content.Context;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Journal de diagnostic de l'enregistrement.
 *
 * Il existe pour une raison precise, apprise a nos depens : quand le
 * declenchement echoue, il echoue SILENCIEUSEMENT. Android refuse de demarrer
 * le service, le recepteur attrape l'exception, et l'utilisateur ne constate
 * qu'une absence — un trajet qui n'est jamais apparu. Sans trace ecrite, il n'y
 * a alors rien a diagnostiquer.
 *
 * Le journal est ecrit au fil de l'eau dans le dossier externe de
 * l'application, donc lisible depuis un gestionnaire de fichiers et
 * partageable, sans quoi il ne servirait a rien.
 */
public final class Diary {

    private static final String FILE_NAME = "journal.txt";
    private static final int MAX_BYTES = 512 * 1024;

    private static final SimpleDateFormat STAMP =
        new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.FRANCE);

    private static final Object LOCK = new Object();

    private Diary() {}

    public static File file(Context context) {
        return new File(Sessions.baseDirectory(context), FILE_NAME);
    }

    public static void log(Context context, String message) {
        synchronized (LOCK) {
            try {
                File journal = file(context);

                // Un journal qui grossit sans fin finirait par remplir le
                // telephone : au-dela d'un demi-megaoctet, on repart a zero en
                // gardant une ligne qui dit ce qui s'est passe.
                if (journal.length() > MAX_BYTES) {
                    write(journal, STAMP.format(new Date()) + "  --- journal tronque ---\n", false);
                }

                write(journal, STAMP.format(new Date()) + "  " + message + "\n", true);
            } catch (IOException ignored) {
                // Un journal illisible ne doit jamais empecher un enregistrement.
            }
        }
    }

    public static String read(Context context) {
        File journal = file(context);
        if (!journal.exists()) return "";

        try {
            byte[] buffer = new byte[(int) journal.length()];
            try (RandomAccessFile in = new RandomAccessFile(journal, "r")) {
                in.readFully(buffer);
            }
            return new String(buffer, StandardCharsets.UTF_8);
        } catch (IOException error) {
            return "Journal illisible : " + error.getMessage();
        }
    }

    public static void clear(Context context) {
        synchronized (LOCK) {
            file(context).delete();
        }
    }

    private static void write(File file, String text, boolean append) throws IOException {
        try (RandomAccessFile out = new RandomAccessFile(file, "rw")) {
            if (append) out.seek(out.length());
            else out.setLength(0);
            out.write(text.getBytes(StandardCharsets.UTF_8));
        }
    }
}

File: android-spike/app/src/main/java/eu/agilmea/spike/Diary.kt
package eu.agilmea.spike

import android.content.Context
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Journal de diagnostic.
 *
 * C'est la piece a conviction de l'experience. Si One UI interrompt
 * l'enregistrement, seul un journal ecrit au fil de l'eau permettra de dire a
 * quelle heure et apres quel evenement — un resume calcule a la fin ne
 * survivrait pas a la mort du processus.
 *
 * Volontairement un simple fichier texte : lisible sans outil, partageable tel
 * quel.
 */
object Diary {

    private const val FILE_NAME = "journal.txt"
    private val stamp = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.FRANCE)
    private val lock = Any()

    fun file(context: Context): File = File(context.getExternalFilesDir(null), FILE_NAME)

    fun log(context: Context, message: String) {
        synchronized(lock) {
            runCatching {
                file(context).appendText("${stamp.format(Date())}  $message\n")
            }
        }
    }

    /** Dernieres lignes, pour l'affichage a l'ecran. */
    fun tail(context: Context, lines: Int): String {
        val f = file(context)
        if (!f.exists()) return "Journal vide."
        return runCatching {
            f.readLines().takeLast(lines).joinToString("\n")
        }.getOrElse { "Journal illisible : ${it.message}" }
    }

    fun clear(context: Context) {
        synchronized(lock) {
            runCatching { file(context).delete() }
        }
    }
}
